Allow intercepting exceptions that are raised in ActionCable command

waymondo opened the pull request:
### Summary

Exceptions raised during command execution on an ActionCable Channel
are logged and suppressed, which does not allow for any sort of error
handling or reporting (via external services such as Bugsnag or Honeybadger).
This adds a simple hook to intercept the rescued exception on the ActionCable Connection.

### Other Information

I had thought that maybe the default implementation of `on_command_execution_exception` should be the current logging behavior, but I opted for it being a no op. I'm happy to change it if that is preferable.

This is my first Rails PR so please let me know if there are any conventions that I should change about this patch.

waymondo commented:
Is there a way I could get a maintainer to weigh in on this PR?

waymondo commented:
Rails bot just reminded me this hasn't been addressed. I still think this is something that should be supported by ActionCable if not this exact PR. It would be great to get a Rails maintainer to weigh in

machty commented:
This is a very useful contribution, could someone please take a look?

jeremy reviewed (commented):
Thanks @waymondo!

Let's use ActiveSupport::Rescuable.

Check out https://github.com/rails/rails/pull/34686/files for an example of the usage style.

waymondo commented:
Thanks @jeremy. That's an obvious improvement in retrospect. Just updated the PR.

jeremy reviewed (commented):
Looking good!

Noting that a Channel rescue_from would handle the exception example in the test case. In this sense, setting a Connection rescue_from is a convenient fallback, in addition to catching other unhandled exceptions.

Let's update the changelog as well!

waymondo commented:
Yeah, I only learned since you linked that PR as an example that the `rescue_from` mechanism was already in place for `Channel`s. This is still helpful for global error reporting, but handling errors in the Channel is probably preferential in most cases. Added a changelog entry, but it might be good to describe these exception handling usages in https://guides.rubyonrails.org/action_cable_overview.html as well. Is there a repo for those, or how are those docs managed?

jeremy commented:
Same git repo! Guides can be updated alongside this PR, if you're interested: https://github.com/rails/rails/blob/master/guides/source/action_cable_overview.md

waymondo committed:
add rescue_with support to ActionCable::Connection::Base

and update ActionCable guide to describe exception handling usage

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# On branch master
# Your branch is behind 'origin/master' by 5 commits, and can be fast-forwarded.
#
# Changes to be committed:
#	modified:   actioncable/CHANGELOG.md
#	modified:   actioncable/lib/action_cable/connection/base.rb
#	modified:   actioncable/lib/action_cable/connection/subscriptions.rb
#	modified:   actioncable/test/connection/subscriptions_test.rb
#	modified:   guides/source/action_cable_overview.md
#

waymondo commented:
Cool, updated the guides and combined everything into a single commit. Let me know if there's anything else you'd like me to adjust.

rmacklin reviewed (commented):
Hey folks, I noticed a typo in the newly added docs, so I've opened https://github.com/rails/rails/pull/38781 to address it.

waymondo commented on guides/source/action_cable_overview.md:
```diff
@@ -128,6 +128,27 @@ can use this approach:
+
+By default, unhandled exceptions are caught and logged to Rails' logger. If you would like to
+globally intercept these exceptions and report them to an external bug tracking service, for
+example, you can do so with `rescue_with`.
```
thanks and sorry!

rmacklin replied:
No problem!

agarcher commented:
I noticed that this implementation doesn't rescue exceptions raised in `subscribed` or `unsubscribed` only in actions. I assume that's not by design because, as far as I can tell, there's no way to recover from exceptions in those flows.
